cloud/vmware/vmware_vm_facts.py

```diff
@@ -51,41 +81,93 @@
+                "ip": [],
             }
         }
 
+        # This may or maynot be present
+        ip = list()
+        ip.append(summary.guest.ipAddress)
+        virtual_machine[name]["ip"] = ip
+        # This may fail, if VM only has one ip
+        for net in vm.guest.net:
+            for iaddr in net.ipConfig.ipAddress :
+                ip.append(iaddr.ipAddress)
+        # Ok, update the ip list
+        virtual_machine[name]["ip"] = list(set(ip))
+    except AttributeError:
+        pass
+
+    return virtual_machine
+
+
+def get_all_virtual_machines(content, folder, recurse):
+    virtual_machines = get_all_objs(content, [vim.VirtualMachine], folder, recurse)
+    _virtual_machines = {}
+
+    for vm in virtual_machines:
+        virtual_machine = get_vm_info(vm)
         _virtual_machines.update(virtual_machine)
     return _virtual_machines
 
 
 def main():
-
     argument_spec = vmware_argument_spec()
-    module = AnsibleModule(argument_spec=argument_spec, supports_check_mode=False)
+    argument_spec.update(dict(datacenter=dict(required=False, type='str'),
+                              cluster=dict(required=False, type='str'),
+                              folder=dict(required=False, type='str'),
+                              recurse=dict(default=True, type='bool')))
+
+    module = AnsibleModule(argument_spec=argument_spec,
+                           mutually_exclusive = (['cluster', 'folder'],),
+                           supports_check_mode=False,)
 
     if not HAS_PYVMOMI:
         module.fail_json(msg='pyvmomi is required for this module')
 
     try:
         content = connect_to_api(module)
-        _virtual_machines = get_all_virtual_machines(content)
+        p = module.params
+        datacenter = p['datacenter']
+        cluster = p['cluster']
+        folder = p['folder']
+        recurse = p['recurse']
+
+        _virtual_machines = {}
+        if datacenter:
+            datacenter = find_datacenter_by_name(content, datacenter)
+            if not datacenter:
+                module.fail_json(msg="datacenter not found")
+
+            # list all VMs in a cluster
+            if cluster:
+                cluster = find_cluster_by_name(content, cluster, datacenter)
+                _virtual_machines = get_all_virtual_machines(content, cluster, recurse)
+            # list all VMs inside a folder aka "Templates and folder"
+            elif folder:
+                folder = find_entity_child_by_path(content, datacenter.vmFolder,folder)
+                _virtual_machines = get_all_virtual_machines(content, folder, recurse)
+            # if none of the above, list all VMs inside the datacenter
+            else:
+                _virtual_machines = get_all_virtual_machines(content, datacenter.vmFolder, recurse)
+        # No datacenter option, list everything. This will be a slow process !
+        else:
+            _virtual_machines = get_all_virtual_machines(content, recurse)
```

Comment:
Hmm - I think this would cause a TypeError because two arguments are being passed to a function requiring three.